addon/services/key-manager.js

```diff
@@ -1,196 +1,159 @@
-import Ember from 'ember';
-import keyCodes from '../utils/key-codes';
-import modifierKeys from '../utils/modifier-keys';
-import modifierKeyCodes from '../utils/modifier-key-codes';
-
-const {
-  $,
-  computed,
+import Service from '@ember/service';
+import { getOwner } from '@ember/application';
+import { assign } from '@ember/polyfills';
+import Macro from '../utils/macro';
+import {
   get,
-  getOwner,
-  run,
+  getProperties,
   set,
   setProperties,
-} = Ember;
-const eventNamespace = 'key-manager';
+} from '@ember/object';
+import { debounce } from '@ember/runloop';
+
 const inputElements = [
-  'input',
-  'textarea',
-  'select',
-  "[contenteditable='true']",
+  'INPUT',
+  'SELECT',
+  'TEXTAREA',
 ];
 
-import Combo from '../utils/combo';
+const isInputElement = (element) => {
+  const isContentEditable = element.isContentEditable;
+  const isInput = inputElements.includes(element.tagName);
 
-export default Ember.Service.extend({
-  clearExecutionKeysLater: null,
-  executionKeyClearInterval: 2000,
-  matchFound: false,
-  uid: 0,
-  altKey: false,
-  ctrlKey: false,
-  metaKey: false,
-  shiftKey: false,
+  return isContentEditable || isInput;
+};
 
-  // config options
-  disableOnInput: false,
+export default Service.extend({
+  isDisabledOnInput: false, // Config option
 
   init() {
-    this._super(...arguments);
-    set(this, 'combos', []);
-    this._resetDownKeys();
-    this._registerConfig();
+    this.keydownMacros = [];
+    this.keyupMacros = [];
+    this._registerConfigOptions();
   },
 
-  register({keys, name, selector=$(document), downCallback, upCallback, priority=0, disableOnInput}) {
-    disableOnInput = disableOnInput || get(this, 'disableOnInput');
-
-    ['up', 'down'].forEach((direction) => {
-      const uid = get(this, 'uid');
-      const callback = direction === 'up' ? upCallback : downCallback;
-      const eventName = `key${direction}.${eventNamespace}.${name}.${uid}`;
-      const combo = Combo.create({
-        callback,
-        direction,
-        eventName,
-        keys,
-        name,
-        selector,
-        priority,
-        uid,
-        disableOnInput,
-      });
+  addMacro() {
+    const macroAttrs = this._mergeConfigDefaults(arguments[0]);
+    const macro = Macro.create();
+    macro.setup(macroAttrs);
 
-      get(this, 'combos').pushObject(combo);
-      selector.on(eventName, {
-        eventName,
-      }, run.bind(this, this.handler));
-      this.incrementProperty('uid');
-    });
-  },
+    const keyEvent = get(macro, 'keyEvent');
+    const macros = get(this, `${keyEvent}Macros`);
 
-  deregister({name}) {
-    const combos = get(this, 'combos');
-    const comboMatches = combos.filterBy('name', name);
+    const element = get(macro, 'element');
+    const hasMacrosWithSameElementAndKeyEvent = this._findMacroWithElementAndKeyEvent(element, keyEvent);
```

Comment:
Nit - less about macro here and more `hasListenerForElementAndKeyEvent`

Reply:
I would also consider moving this to a separate method that accepts just the element and keyEvent as params as it is separate to registering the callback function for a macro.